File: my-code/53/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/05/28 17: 52
# @Author   : Ranshi
# @File     : 53. 最大子序和.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res, ans = -float("inf"), 0
        for x in nums:
            ans += x
            res = max(ans, res)
            ans = max(ans, 0)
        return int(res)


if __name__ == "__main__":
    s = Solution()
    print(s.maxSubArray(nums=[-2]))

File: my-code/53/main.go
package main

import "fmt"

// @Time     : 2021/7/22 21:09
// @Author   : Ranshi
// @File     : main.go

func maxSubArray(nums []int) int {
	idx, res := -1, nums[0]
	for i := range nums {
		if idx < 0 {
			idx = nums[i]
		} else {
			idx += nums[i]
		}
		if res < idx {
			res = idx
		}
	}
	return res
}

func main() {
	fmt.Println(maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}
